benchmarks/nel/scripts/datasets/dataset.py

```diff
@@ -0,0 +1,301 @@
+""" Dataset class. """
+import abc
+import importlib
+import inspect
+import os
+import pickle
+from collections import defaultdict
+from pathlib import Path
+from typing import Tuple, Set, List, Optional, TypeVar, Type, Dict
+
+import numpy
+import spacy
+import tqdm
+import yaml
+from spacy import Language
+from spacy.kb import KnowledgeBase
+from spacy.tokens import Doc, DocBin
+from spacy.training import Example
+
+from . import evaluation
+from .utils import ENTITIES_TYPE, ANNOTATIONS_TYPE
+
+DatasetType = TypeVar('DatasetType', bound='Dataset')
+
+
+class Dataset(abc.ABC):
+    """ Base class for all datasets used in this benchmark. """
+
+    KB_VECTOR_LENGTH = 300
+
+    def __init__(self):
+        """ Initializes new Dataset.
+        """
+
+        self._paths = self.assemble_paths(self.dataset_id)
+
+        with open(self._paths["root"] / "configs" / "datasets.yml", "r") as stream:
+            self._options = yaml.safe_load(stream)[self.dataset_id]
+
+        self._entities: Optional[ENTITIES_TYPE] = None
+        self._failed_entity_lookups: Optional[Set[str]] = None
+        self._annotations: Optional[ANNOTATIONS_TYPE] = None
+        self._kb: Optional[KnowledgeBase] = None
+        self._nlp_base: Optional[Language] = None
+        self._nlp_best: Optional[Language] = None
+        self._annotated_docs: Optional[List[Doc]] = None
+
+    @staticmethod
+    def assemble_paths(dataset_id: str) -> Dict[str, Path]:
+        """ Assemble paths w.r.t. dataset ID.
+        dataset_id (str): Dataset ID.
+        RETURNS (Dict[str, Path]): Dictionary with internal resource name to path.
+        """
+
+        root_path = Path(os.path.abspath(__file__)).parent.parent.parent
+        assets_path = root_path / "assets" / dataset_id
+
+        return {
+            "root": root_path,
+            "assets": assets_path,
+            "nlp_base": root_path / "temp" / f"{dataset_id}.nlp",
+            "nlp_best": root_path / "training" / dataset_id / "model-best",
+            "kb": root_path / "temp" / f"{dataset_id}.kb",
```

Comment:
This might just be me, but I find dots in folder names pretty confusing - maybe let's use a dash or underscore instead?

Reply:
Sure!